### PR TITLE
Apib attributes

### PR DESCRIPTION
**Pull request by rhazdon**

### CODE REVIEW

File: lib/blue_bird/json_data.ex [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  ## Example
+
+        iex> JSONData.type("string")
+        "string"
+        iex> JSONData.type(:yes)
+        "string"
+        iex> JSONData.type(1)
+        "number"
+  """
+  @spec type(Any) :: String.t
```
Comment (woylie) [inline lib/blue_bird/json_data.ex]: Should be `type(any)`.

Apart from that, I'm getting these dialyzer warnings:

```
:0: Unknown function 'Elixir.BlueBird.JSONData.Function':'__impl__'/1
:0: Unknown function 'Elixir.BlueBird.JSONData.PID':'__impl__'/1
:0: Unknown function 'Elixir.BlueBird.JSONData.Port':'__impl__'/1
:0: Unknown function 'Elixir.BlueBird.JSONData.Reference':'__impl__'/1
```

I tried setting `@fallback_to_any true`, but that didn't help. So either we have to add useless implementations for Function, PID, Port, and Reference, or we have to configure Dialyxir to ignore these warnings.

File: lib/blue_bird/writer/blueprint.ex [hunk 2]
```diff
@@ -234,6 +236,38 @@ defmodule BlueBird.Writer.Blueprint do
...
+      print_attribute_name_and_type(key, value)
+    end
+  end
+
+  defp print_attribute({key, value}), do: print_attribute_name_and_type(key, value)
+
+  defp print_attribute_name_and_type(key, value), do: "+ #{key} (#{JSONData.type(value)})\n"
+
+  defp walk_through_attributes(attributes, indent \\ 8) do
+    (attributes |> Enum.map_join(&(print_attribute(&1))) |> indent(indent))
```
Comment (woylie) [inline lib/blue_bird/writer/blueprint.ex]: Too many brackets.

File: lib/blue_bird/writer/blueprint.ex [hunk 3]
```diff
@@ -234,6 +236,38 @@ defmodule BlueBird.Writer.Blueprint do
...
+  def print_attributes(attributes) do
+    "+ Attributes (object)\n\n"
+    <> walk_through_attributes(attributes)
+  end
+
+  defp print_attribute({key, value}) when is_map(value) do
+    print_attribute_name_and_type(key, value)
+    <> walk_through_attributes(value, 4)
+  end
+
```
Comment (woylie) [inline lib/blue_bird/writer/blueprint.ex]: Too many line breaks (here and after the function).

File: lib/blue_bird/writer/blueprint.ex [hunk 4]
```diff
@@ -234,6 +236,38 @@ defmodule BlueBird.Writer.Blueprint do
...
+    if is_map(List.first(value)) do
+      print_attribute_name_and_type(key, value)
+      <> ("+ (object)\n" |> indent(4))
+      <> walk_through_attributes(List.first(value))
+    else
+      print_attribute_name_and_type(key, value)
+    end
+  end
+
+  defp print_attribute({key, value}), do: print_attribute_name_and_type(key, value)
```
Comment (woylie) [inline lib/blue_bird/writer/blueprint.ex]: Too long lines.

File: test/generator_test.exs [hunk 5]
```diff
@@ -230,7 +256,31 @@ defmodule BlueBird.Test.GeneratorTest do
 
       assert request.query_params == %{"s" => "poodle"}
       assert request.path_params == %{"id" => "137"}
-      assert request.body_params == %{"betty" => "white"}
+      assert request.body_params == %{
```
Comment (woylie) [inline test/generator_test.exs]: What do we gain from this? The generator only sets `request.body_params` to `conn.body_params`. So we only have to test whether the field was set as expected, but we don't have to test whether Plug is able to decode complex json objects.

File: test/support/controller/undocumented_controller.ex [hunk 6]
```diff
@@ -1,4 +1,4 @@
-defmodule BlueBird.Test.Support.TestControllerUndocumented do
+defmodule BlueBird.Test.Support.TestUndocumentedController do
```
Comment (woylie) [inline test/support/controller/undocumented_controller.ex]: `BlueBird.Test.Support.TestController`, `BlueBird.Test.Support.TestNamedController` and `BlueBird.Test.Support.TestUndocumentedController` is a bit redundant. Can we rename them to `BlueBird.Test.Support.Controller` etc.?

File: lib/blue_bird/writer/blueprint.ex (hunk 4 above)
Comment (woylie) [reply lib/blue_bird/writer/blueprint.ex]: Can you check the credo configuration? It doesn't emit a warning here.

Comment (woylie) [reply lib/blue_bird/writer/blueprint.ex]: Also, there's a credo warning for `generator.ex` line 176, which you didn't change, so I wonder why it went through Travis without notice.

File: test/writer/blueprint_test.exs [hunk 7]
```diff
@@ -56,6 +82,34 @@ defmodule BlueBird.Test.Writer.BlueprintTest do
     end
   end
 
+  describe "print_attributes/1" do
+    test "prints empty attributes correctly" do
+      assert print_attributes(%{}) == ""
+    end
+
+    test "prints attributes correctly" do
+      assert print_attributes(@complex_body) ==
```
Comment (woylie) [inline test/writer/blueprint_test.exs]: Can we put the first `"""` in the assert-line, and then indent the block beginning with `+ Attributes` with two spaces?

**Review by woylie — CHANGES_REQUESTED** [review]

File: lib/blue_bird/json_data.ex (hunk 1 above)
Comment (rhazdon) [reply lib/blue_bird/json_data.ex]: Thank you, I will fix the typo.

Regardless the `Unknown function` I would prefer to add the missing implementations. The return value should be `null`.

File: lib/blue_bird/writer/blueprint.ex (hunk 4 above)
Comment (rhazdon) [reply lib/blue_bird/writer/blueprint.ex]: I do not have any credo warnings at all.... I'm always using credo with `--strict`...

File: test/writer/blueprint_test.exs (hunk 7 above)
Comment (rhazdon) [reply test/writer/blueprint_test.exs]: Of course. But I will change it then on every other place, too. Consistency, you know. :)

Comment (woylie) [reply test/writer/blueprint_test.exs]: Thanks 👍

File: test/generator_test.exs (hunk 5 above)
Comment (rhazdon) [reply test/generator_test.exs]: Mhh, you might be right... this test is maybe not that useful. In fact the function itself is tested and it is not necessary anymore to do that here. I will change that.

File: test/support/controller/undocumented_controller.ex (hunk 6 above)
Comment (rhazdon) [reply test/support/controller/undocumented_controller.ex]: Yes... sure.

File: lib/blue_bird/writer/blueprint.ex [hunk 8]
```diff
@@ -245,6 +245,7 @@ defmodule BlueBird.Writer.Blueprint do
     <> walk_through_attributes(attributes)
   end
 
+  @spec print_attribute(Tuple.t) :: String.t
```
Comment (woylie) [inline lib/blue_bird/writer/blueprint.ex]: This has to be `@spec print_attribute({String.t, any}) :: String.t`.

File: lib/blue_bird/writer/blueprint.ex [hunk 9]
```diff
@@ -234,6 +236,45 @@ defmodule BlueBird.Writer.Blueprint do
...
+  ## Attributes
+
+  @spec print_attributes(map) :: String.t
+  def print_attributes(attributes) when attributes == %{}, do: ""
+  def print_attributes(attributes) do
+    "+ Attributes (object)\n\n"
+    <> walk_through_attributes(attributes)
+  end
+
+  @spec print_attribute(Tuple.t) :: String.t
```
Comment (woylie) [inline lib/blue_bird/writer/blueprint.ex]: This has to be `@spec print_attribute({String.t, any}) :: String.t`.

**Review by woylie — CHANGES_REQUESTED** [review]

**Review by woylie — APPROVED** [review]

**Pull request merged**